### DISCUSSION TITLE
RuntimeError: applying transform RandCropByPosNegLabeld

### DISCUSSION THREAD

**Discussion by dianemarquette** [Project-MONAI/MONAI#1811] [Q&A] [answered]
Hi :)

Thank you for the great package. The detailed documentation is also very much appreciated ! 

I have successfully run the Spleen 3D Segmentation notebook with the provided data. I'm now trying to get it to work with my custom dataset (CT scans of hands in NIfTI format with 3 classes: background, trapezium & metacarpial bone of the thumb). 

**My problem**
Whereas the first cells execute smoothly, the "typical Pytorch training process" cell gives me the following error:

-----
epoch 1/600
1/11, train_loss: 0.8117
2/11, train_loss: 0.8081
3/11, train_loss: 0.8139

ValueError                                Traceback (most recent call last)
~\anaconda3\envs\kerimedical\lib\site-packages\monai\transforms\utils.py in apply_transform(transform, data, map_items)
    360             return [transform(item) for item in data]
--> 361         return transform(data)
    362     except Exception as e:

~\anaconda3\envs\kerimedical\lib\site-packages\monai\transforms\croppad\dictionary.py in __call__(self, data)
    534 
--> 535         self.randomize(label, fg_indices, bg_indices, image)
    536         assert isinstance(self.spatial_size, tuple)

~\anaconda3\envs\kerimedical\lib\site-packages\monai\transforms\croppad\dictionary.py in randomize(self, label, fg_indices, bg_indices, image)
    519         if fg_indices is None or bg_indices is None:
--> 520             fg_indices_, bg_indices_ = map_binary_to_indices(label, image, self.image_threshold)
    521         else:

~\anaconda3\envs\kerimedical\lib\site-packages\monai\transforms\utils.py in map_binary_to_indices(label, image, image_threshold)
    215         img_flat = np.any(image > image_threshold, axis=0).ravel()
--> 216         bg_indices = np.nonzero(np.logical_and(img_flat, ~label_flat))[0]
    217     else:

ValueError: operands could not be broadcast together with shapes (778998,) (0,) 

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
<ipython-input-68-449a99c20834> in <module>
     14     epoch_loss = 0
     15     step = 0
---> 16     for batch_data in train_loader:
     17         step += 1
     18         inputs, labels = (

~\anaconda3\envs\kerimedical\lib\site-packages\torch\utils\data\dataloader.py in __next__(self)
    515             if self._sampler_iter is None:
    516                 self._reset()
--> 517             data = self._next_data()
    518             self._num_yielded += 1
    519             if self._dataset_kind == _DatasetKind.Iterable and \

~\anaconda3\envs\kerimedical\lib\site-packages\torch\utils\data\dataloader.py in _next_data(self)
    555     def _next_data(self):
    556         index = self._next_index()  # may raise StopIteration
--> 557         data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
    558         if self._pin_memory:
    559             data = _utils.pin_memory.pin_memory(data)

~\anaconda3\envs\kerimedical\lib\site-packages\torch\utils\data\_utils\fetch.py in fetch(self, possibly_batched_index)
     42     def fetch(self, possibly_batched_index):
     43         if self.auto_collation:
---> 44             data = [self.dataset[idx] for idx in possibly_batched_index]
     45         else:
     46             data = self.dataset[possibly_batched_index]

~\anaconda3\envs\kerimedical\lib\site-packages\torch\utils\data\_utils\fetch.py in <listcomp>(.0)
     42     def fetch(self, possibly_batched_index):
     43         if self.auto_collation:
---> 44             data = [self.dataset[idx] for idx in possibly_batched_index]
     45         else:
     46             data = self.dataset[possibly_batched_index]

~\anaconda3\envs\kerimedical\lib\site-packages\monai\data\dataset.py in __getitem__(self, index)
    531             if start_run or isinstance(_transform, Randomizable) or not isinstance(_transform, Transform):
    532                 start_run = True
--> 533                 data = apply_transform(_transform, data)
    534         return data
    535 

~\anaconda3\envs\kerimedical\lib\site-packages\monai\transforms\utils.py in apply_transform(transform, data, map_items)
    361         return transform(data)
    362     except Exception as e:
--> 363         raise RuntimeError(f"applying transform {transform}") from e
    364 
    365 

RuntimeError: applying transform <monai.transforms.croppad.dictionary.RandCropByPosNegLabeld object at 0x00000277E4C14F10>

-----

**What didn't solve the problem**
I've tried to play with the `spatial_size` of the `RandCropByPosNegLabeld` transform but it didn't help. I don't understand what the "ValueError: operands could not be broadcast together with shapes (778998,) (0,)" implies. 

Any help or suggestion would be greatly appreciated :)
Thanks for reading!

**Comment by Nic-Ma** [comment]
Hi @dianemarquette ,

I feel it's something related to numpy version. Maybe you can update your numpy and try again?
@wyli , dou you have any more suggestion?

Thanks.

**Reply by wyli** [reply]
could you please share the input parameters? I couldn't replicate this issue -- looks like `label_flat`'s shape is `(0,)`

**Reply by dianemarquette** [reply]
Thank you for your suggestions.

@Nic-Ma I'm already using the latest version of numpy (1.20.1). So, it's probably not due to this.

@wyli I'm using the tutorial for the spleen 3D segmentation. I didn't change a lot of code. The only two cells I changed are the ones defining the transforms and the model (modifications are in bold).

### Transform cell

train_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        AddChanneld(keys=["image", "label"]),
        Spacingd(keys=["image", "label"], pixdim=(
            **0.4, 0.4, 0.625**), mode=("bilinear", "nearest")), # 1.5, 1.5, 2
        Orientationd(keys=["image", "label"], axcodes="RAS"),
        **DataStatsd(keys=['image', 'label'])**, # added
        ScaleIntensityRanged(
            keys=["image"], **a_min=200, a_max=2000**,
            b_min=0.0, b_max=1.0, clip=True,
        ),
        CropForegroundd(keys=["image", "label"], source_key="image"),
        RandCropByPosNegLabeld(
            keys=["image", "label"],
            label_key="label",
            # spatial_size=(96, 96, 96),
            **spatial_size=(16,16,16)**,
            pos=1,
            neg=1,
            num_samples=4,
            image_key="image",
            image_threshold=0,
        ),
        # user can also add other random transforms
        # RandAffined(
        #     keys=['image', 'label'],
        #     mode=('bilinear', 'nearest'),
        #     prob=1.0, spatial_size=(96, 96, 96),
        #     rotate_range=(0, 0, np.pi/15),
        #     scale_range=(0.1, 0.1, 0.1)),
        ToTensord(keys=["image", "label"]),
    ]
)
val_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        AddChanneld(keys=["image", "label"]),
        Spacingd(keys=["image", "label"], pixdim=(
            **0.4, 0.4, 0.625**), mode=("bilinear", "nearest")), # 1.5, 1.5, 2
        Orientationd(keys=["image", "label"], axcodes="RAS"),
        **DataStatsd(keys=['image', 'label']),** # added
        ScaleIntensityRanged(
            keys=["image"], **a_min=200, a_max=2000**,
            b_min=0.0, b_max=1.0, clip=True,
        ),
        CropForegroundd(keys=["image", "label"], source_key="image"),
        ToTensord(keys=["image", "label"]),
    ]
)

### Model cell 

device = torch.device("cuda:0")
model = UNet(
    dimensions=3,
    in_channels=1,
    **out_channels=3, # 3 classes = background + trapezium + metacarpial**
    channels=(16, 32, 64, 128, 256),
    strides=(2, 2, 2, 2),
    num_res_units=2,
    norm=Norm.BATCH,
).to(device)
loss_function = DiceLoss(to_onehot_y=True, softmax=True)
optimizer = torch.optim.Adam(model.parameters(), 1e-4)

The output of the DataLoader looks ok to me:
![check_data output](https://user-images.githubusercontent.com/60931186/112622434-507c6680-8e2b-11eb-89a5-fe9aee90e477.png)

However, I reckon that the fact that `label_flat`'s shape is `(0,)` is strange. I still didn't find a solution, so any ideas would be extremely helpful !

Thank you for your time.

**Comment by dianemarquette** [comment, marked as answer]
I ended up understanding the cause of this error. I'm writing it here in case it can help someone else.

The problem didn't come from the code. It was the dataset. The labels of two CT scans must have been incorrectly exported. As a result, the dimensions of the labels didn't match the dimensions of the scans (for scans #6 and #10).

![Load dataset (0 4, 0 4, 0 625)](https://user-images.githubusercontent.com/60931186/112957109-6c407f00-9141-11eb-9bd3-75d9d11f4d09.png)

Deleting these mislabeled scans from my dataset solved the problem :). 

Cheers,

Diane
